src/librustc_metadata/cstore.rs

```diff
@@ -218,6 +222,23 @@ impl CStore {
         self.used_libraries.borrow_mut().push(lib);
     }
 
+    // Update kind and, optionally, the name of all native libaries (there may be more than one)
+    // with the specified name.
+    pub fn update_used_library(&self, name: &str, new_name: Option<&str>,
+                               new_kind: NativeLibraryKind) -> bool {
+        let mut found = false;
+        for item in self.used_libraries.borrow_mut().iter_mut() {
+            if item.name == name {
+                item.kind = new_kind;
+                if let Some(new_name) = new_name {
+                    item.name = Symbol::intern(new_name);
```

Comment:
Not sure what you mean by "multiple indirections"...  Is this about the fact that there may be several `NativeLibrary` entries for the same library?   I don't think we can just merge them,- because library ordering on linker's command line is important.